### PR TITLE
Add `podman container cleanup` to CLI

### PR DESCRIPTION
**Pull request by rhatdan**
When we run containers in detach mode, nothing cleans up the network stack or
the mount points.  This patch will tell conmon to execute the cleanup code when
the container exits.

It can also be called to attempt to cleanup previously running containers.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
@mheon This is the first pass at the cleanup code to be called by conmon when detatched containers exit.  Probably need to store these commands in the database, since you could create a container and later start it, do you agree?

**Comment by rhatdan** [comment]
@giuseppe PTAL

**Comment by mheon** [comment]
@rhatdan Yes, these need to be in the database, definitely.
The difficulty here is going to be if someone runs `podman` with a non-standard store (changing storage driver, for example). We need to be smart enough to put that in the cleanup command.

File: docs/podman-containers-cleanup.1.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+## NAME
+podman\-container\-cleanup - Cleanup Container storage and networks
+
+## SYNOPSIS
+**podman** **container-cleanup** [*options* [...]] container
+
+## DESCRIPTION
+`podman container cleanup` will cleanup one or more containers from the host.  The container name or ID can be used.  This does not remove containers.  Running containers will not be cleaned up.
+Some times containers mount points and network stacks can remain if the podman command was killed or the container ran as a daemon.  This command will run the finall cleanup
```
Comment (giuseppe) [inline docs/podman-containers-cleanup.1.md]: typo here, final cleanup.

**Comment by rhatdan** [comment]
@mheon @giuseppe @umohnani8 PTAL

File: pkg/spec/createconfig.go [hunk 2]
```diff
@@ -315,8 +315,21 @@ func (c *CreateConfig) GetTmpfsMounts() []spec.Mount {
 	return m
 }
 
+func createExitCommand(runtime *libpod.Runtime) []string {
+	config := runtime.GetConfig()
+
+	command := []string{os.Args[0],
```
Comment (mheon) [inline pkg/spec/createconfig.go]: Should we be including anything else from the storage config in here

File: libpod/options.go [hunk 3]
```diff
@@ -485,6 +485,18 @@ func WithIDMappings(idmappings storage.IDMappingOptions) CtrCreateOption {
...
 }
 
+// WithExitCommand sets the ExitCommand for the container, appending on the ctr.ID() to the end
+func WithExitCommand(exitCommand []string) CtrCreateOption {
+	return func(ctr *Container) error {
+		if ctr.valid {
+			return ErrCtrFinalized
+		}
+
+		ctr.config.ExitCommand = append(exitCommand, ctr.ID())
```
Comment (mheon) [inline libpod/options.go]: We should make this conditional, some applications might not need it. Add a `appendContainerId bool` argument to `WithExitCommand` maybe?

File: pkg/spec/createconfig.go [hunk 4]
```diff
@@ -433,6 +446,9 @@ func (c *CreateConfig) GetContainerCreateOptions() ([]libpod.CtrCreateOption, er
 	if c.CgroupParent != "" {
 		options = append(options, libpod.WithCgroupParent(c.CgroupParent))
 	}
+	if c.Detach {
+		options = append(options, libpod.WithExitCommand(createExitCommand(runtime)))
```
Comment (mheon) [inline pkg/spec/createconfig.go]: We should probably do this unconditionally, and remove the calls to `Cleanup()` in `podman run`

File: libpod/container_api.go [hunk 5]
```diff
@@ -753,3 +753,8 @@ func (c *Container) RestartWithTimeout(ctx context.Context, timeout uint) (err e
 
 	return c.start()
 }
+
+// SetExitCommand sets the command to be executed by conmon when the container exits
+func (c *Container) SetExitCommand(command []string) {
```
Comment (mheon) [inline libpod/container_api.go]: Can't do this, we never overwrite config in the DB

Comment (rhatdan) [reply libpod/container_api.go]: Removed that was cruft from a previous attempt.

File: pkg/spec/createconfig.go (hunk 4 above)
Comment (rhatdan) [reply pkg/spec/createconfig.go]: I would prefer to wait for this, make sure this is working well before we make it the default.

File: pkg/spec/createconfig.go (hunk 2 above)
Comment (rhatdan) [reply pkg/spec/createconfig.go]: I was looking for stuff that would effect the cleanup code, and did not see anything.  
log-level?
Here are the global options:
```
"cgroup-manager",     N Not running containers
"cni-config-dir",           N Not running containers
"config, c",                  M But i don't think so
"conmon",                  N Not running containers
"cpu-profile",              N Not running containers
"default-mounts-file",  N Not running containers
"hooks-dir-path",         N: Not running containers.
"log-level",                  M: Maybe
"storage-driver, s",       Y: We have this one.
"storage-opt",              M: Maybe
"root",                          Y: We have this one.
"tmpdir",                      N: Don't think so
"runroot"                     N: We have this one.
"runtime"                     N: We are not running containers
"storage-driver, s"       Y: Have this one.
"storage-opt"               M: I don't think we need this.
```

Comment (mheon) [reply pkg/spec/createconfig.go]: We need `cgroup-manager` set to clean up CGroups. Might be a good idea to set tmpdir, too.

Comment (rhatdan) [reply pkg/spec/createconfig.go]: Added

File: libpod/options.go (hunk 3 above)
Comment (rhatdan) [reply libpod/options.go]: Why would I call WithExitCommand if I did not need it?

File: docs/podman-containers-cleanup.1.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
+% podman-container-cleanup "1"
+
+## NAME
+podman\-container\-cleanup - Cleanup Container storage and networks
+
+## SYNOPSIS
+**podman** **container-cleanup** [*options* [...]] container
+
+## DESCRIPTION
+`podman container cleanup` will cleanup one or more containers from the host.  The container name or ID can be used.  This does not remove containers.  Running containers will not be cleaned up.
```
Comment (TomSweeneyRedHat) [inline docs/podman-containers-cleanup.1.md]: suggest "remove containers" to "remove the container"

Comment (TomSweeneyRedHat) [reply docs/podman-containers-cleanup.1.md]: oops, forgot the --all flag..  suggest adding 'the' only then.  "remove the containers".

File: docs/podman-containers-cleanup.1.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
+% podman-container-cleanup "1"
+
+## NAME
+podman\-container\-cleanup - Cleanup Container storage and networks
+
+## SYNOPSIS
+**podman** **container-cleanup** [*options* [...]] container
+
+## DESCRIPTION
```
Comment (TomSweeneyRedHat) [inline docs/podman-containers-cleanup.1.md]: After reading this, I'm not sure what "clean up" is meant to do.  Can you add a few details?  Perhaps "Cleans the container by removing all mountpoints and networks."

**Comment by rhatdan** [comment]
@TomSweeneyRedHat Fixed up the man page.

**Comment by rhatdan** [comment]
@mheon @giuseppe @TomSweeneyRedHat PTAL

File: docs/podman-containers-cleanup.1.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+## OPTIONS
+
+**--all, a**
+
+Cleanup all containers.
+
+**--latest, -l**
+Instead of providing the container name or ID, use the last created container. If you use methods other than Podman
+to run containers such as CRI-O, the last started container could be from either of those methods.
+## EXAMPLE
```
Comment (TomSweeneyRedHat) [inline docs/podman-containers-cleanup.1.md]: nit, should there be a blank line before this `#EXAMPLE`?

File: docs/podman-containers-cleanup.1.md (hunk 7 above)
Comment (TomSweeneyRedHat) [reply docs/podman-containers-cleanup.1.md]: I like the added verbiage, tyvm!

**Comment by TomSweeneyRedHat** [comment]
One small nit on the man page, otherwise LGTM.  Would like a head nod from @mheon.

File: cmd/podman/cleanup.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+	"github.com/projectatomic/libpod/cmd/podman/libpodruntime"
+	"github.com/projectatomic/libpod/libpod"
+	"github.com/urfave/cli"
+)
+
+var (
+	cleanupFlags = []cli.Flag{
+		cli.BoolFlag{
+			Name:  "all, a",
+			Usage: "Remove all containers",
```
Comment (mheon) [inline cmd/podman/cleanup.go]: Remove all?

Also, does this flag even make sense? If we ever end up in a situation where we need to manually clean up every container, something has gone seriously wrong.

File: libpod/options.go (hunk 3 above)
Comment (mheon) [reply libpod/options.go]: I can see people wanting to call non-Podman programs after a container exits - maybe sendmail to throw out a "this container has died" message?

File: cmd/podman/cleanup.go (hunk 9 above)
Comment (TomSweeneyRedHat) [reply cmd/podman/cleanup.go]: @mheon that's a good point, it really only cleans all exited containers and probably should say that if we're keeping the --all param.

File: cmd/podman/cleanup.go [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+		Usage:       "Cleanup network and mountpoints of one or more containers",
+		Description: cleanupDescription,
+		Flags:       cleanupFlags,
+		Action:      cleanupCmd,
+		ArgsUsage:   "CONTAINER-NAME [CONTAINER-NAME ...]",
+	}
+)
+
+func cleanupCmd(c *cli.Context) error {
+	runtime, err := libpodruntime.GetRuntime(c)
```
Comment (mheon) [inline cmd/podman/cleanup.go]: Don't we want a `validateFlags()` in here?

File: cmd/podman/cleanup.go [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+		cli.BoolFlag{
+			Name:  "all, a",
+			Usage: "Remove all containers",
+		},
+		LatestFlag,
+	}
+	cleanupDescription = `
+   podman container cleanup
+
+   Cleanups one or more stopped containers.  The container name or ID can be used.
```
Comment (mheon) [inline cmd/podman/cleanup.go]: Cleanups -> Cleans up

File: cmd/podman/cleanup.go (hunk 9 above)
Comment (rhatdan) [reply cmd/podman/cleanup.go]: Cut and paste, fixed.

**Comment by rhatdan** [comment]
Sadly this is still not working correctly.  IE Conmon is not actually executing the cleanup.  Continuing to investigate.

**Comment by giuseppe** [comment]
this patch in conmon is needed: https://github.com/kubernetes-incubator/cri-o/pull/1641

I tested this PR with the patched conmon and it works well for me

File: cmd/podman/cleanup.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+		cli.BoolFlag{
+			Name:  "all, a",
+			Usage: "Cleans up all containers",
+		},
+		LatestFlag,
+	}
+	cleanupDescription = `
+   podman container cleanup
+
+   Cleans up mount points and network stacks on one or more containers from the host.  The container name or ID can be used. This command is used internally when running containers, but can  also be used if container cleanup has failed when a container exits.
```
Comment (wking) [inline cmd/podman/cleanup.go]: Looks like you have two spaces in `can  also`.

File: docs/podman-container.1.md [hunk 13]
```diff
@@ -14,6 +14,7 @@ The container command allows you to manage containers
 | Command  | Man Page                                            | Description                                                                  |
 | -------  | --------------------------------------------------- | ---------------------------------------------------------------------------- |
 | attach   | [podman-attach(1)](podman-attach.1.md)              | Attach to a running container.                                               |
+| cleanup  | [podman-container-cleanup(1)](podman-container-cleanup.1.md)    | Cleanup containers network and mountpoints.                               |
```
Comment (wking) [inline docs/podman-container.1.md]: This man-page URL includes `-container`, which is missing from the surrounding entries.  Do those files need to be renamed to include it?

File: docs/podman-containers-cleanup.1.md [hunk 14]
```diff
@@ -0,0 +1,38 @@
+% podman-container-cleanup "1"
+
+## NAME
+podman\-container\-cleanup - Cleanup Container storage and networks
+
+## SYNOPSIS
+**podman** **container-cleanup** [*options* [...]] container
```
Comment (wking) [inline docs/podman-containers-cleanup.1.md]: Shouldn't this be:

```markdown
**podman container cleanup** [*options* [...]] *container*
```

File: docs/podman-containers-cleanup.1.md [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+## DESCRIPTION
+`podman container cleanup` cleans up exited containers by removing all mountpoints and network configuration from the host.  The container name or ID can be used.  The cleanup command does not remove the containers.  Running containers will not be cleaned up.
+Some times containers mount points and network stacks can remain if the podman command was killed or the container ran in daemon mode.  This command is automatically executed when you run containers in daemon mode by the conmon process when the container exits.
```
Comment (wking) [inline docs/podman-containers-cleanup.1.md]: `containers mount points` -> `container's mount points`.  And `Some times` -> `Sometimes`?

File: docs/podman-containers-cleanup.1.md [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+podman container cleanup mywebserver
+
+podman container cleanup mywebserver myflaskserver 860a4b23
+
+podman container cleanup -f 860a4b23
+
+podman container-cleanup -a
+
+podman container cleanup --latest
```
Comment (wking) [inline docs/podman-containers-cleanup.1.md]: I think you want fencing backticks or four-space indents for all of these examples.  And also possibly English explaining the point you're illustrating with each.

File: docs/podman-containers-cleanup.1.md [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+**--latest, -l**
+Instead of providing the container name or ID, use the last created container. If you use methods other than Podman
+to run containers such as CRI-O, the last started container could be from either of those methods.
+## EXAMPLE
+
+podman container cleanup mywebserver
+
+podman container cleanup mywebserver myflaskserver 860a4b23
+
+podman container cleanup -f 860a4b23
```
Comment (wking) [inline docs/podman-containers-cleanup.1.md]: You don't document an `-f` above.

File: libpod/options.go (hunk 3 above)
Comment (wking) [reply libpod/options.go]: > I can see people wanting to call non-Podman programs after a container exits...

How about recycling the OCI hook structure for these post-exit commands?  Have `conmon` cann a `podman` subcommand with instructions for finding the post-exit hook JSON, and make `cleanup` one of those entries.  Or wecan address all of that in follow-up work.

Comment (mheon) [reply libpod/options.go]: I think having a dedicated "postexit" API we call would be a good idea (clean up, store the exit code in the DB, and call an OCI extension hook for postexit). I see that as conflicting with this work to a degree - we wouldn't want to let the postexit command be manually specifed through the API, we'd just have them tell us where the executable lives so we can invoke Podman or CRI-O again and start the postexit call.

Comment (rhatdan) [reply libpod/options.go]: Lets do that later.  For now lets get this merged so we stop leaking mount points all over the place, and we can work on adding advanced features like a poststop hook in later PRs.

File: docs/podman-containers-cleanup.1.md (hunk 8 above)
Comment (rhatdan) [reply docs/podman-containers-cleanup.1.md]: Fixed

**Comment by mheon** [comment]
@rhatdan Can we get a test that verifies that mounts are cleaned up with detached containers after this?

File: cmd/podman/cleanup.go (hunk 12 above)
Comment (rhatdan) [reply cmd/podman/cleanup.go]: Fixed

File: docs/podman-container.1.md (hunk 13 above)
Comment (rhatdan) [reply docs/podman-container.1.md]: No, podman container attach == podman attach, it is just a different way of executing the same code.

BTW
docker container attach == docker attach also

File: docs/podman-containers-cleanup.1.md (hunk 14 above)
Comment (rhatdan) [reply docs/podman-containers-cleanup.1.md]: Fixed

File: docs/podman-containers-cleanup.1.md (hunk 15 above)
Comment (rhatdan) [reply docs/podman-containers-cleanup.1.md]: fixed

File: docs/podman-container.1.md (hunk 13 above)
Comment (wking) [reply docs/podman-container.1.md]: > No, podman container attach == podman attach...

And similarly for all of the other commands here?  It looks like `refresh` (at `podman-container-refresh.1.md`) is the only existing `container` subcommand that has a `container`-specific man-page.

File: docs/podman-containers-cleanup.1.md (hunk 16 above)
Comment (rhatdan) [reply docs/podman-containers-cleanup.1.md]: fixed

File: docs/podman-containers-cleanup.1.md (hunk 17 above)
Comment (rhatdan) [reply docs/podman-containers-cleanup.1.md]: Removed

**Comment by rhatdan** [comment]
Yes I will add tests before merging.  Need updated conmon to make this work at all.

File: docs/podman-container.1.md (hunk 13 above)
Comment (rhatdan) [reply docs/podman-container.1.md]: Yes we could create some links for the others, so that `man podman container attach` would show you the `podman attach` man page.

File: cmd/podman/cleanup.go [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+	if err != nil {
+		return errors.Wrapf(err, "could not get runtime")
+	}
+	defer runtime.Shutdown(false)
+
+	args := c.Args()
+	if c.Bool("latest") && c.Bool("all") {
+		return errors.Errorf("--all and --latest cannot be used together")
+	}
+	if len(args) == 0 && !c.Bool("all") && !c.Bool("latest") {
```
Comment (wking) [inline cmd/podman/cleanup.go]: There are some cases you aren't covering here.  For example, if both `--latest` and an explicit container ID are given, you are using the latest container and ignoring the explicitly-given ID.  Perhaps something like:

```go
if c.Bool("all") {
  if c.Bool("lastest") {
    return errors.Errorf("--all and --latest cannot be used together")
  } else if len(args) != 0 {
    return errors.Errorf("--all and explicit container IDs cannot be used together")
  }
} else if c.Bool("latest") {
  if len(args) != 0 {
    return errors.Errorf("--latest and explicit container IDs cannot be used together")
  }
} else if len(args) == 0 {
  return errors.Errorf("you must provide at least one container name or id")
}
```

Also, there's no formatting going on here, so [`Errorf` from the GitHub package][1] seems like overkill.  Can we use [the stdlib's `errors.New`][2]?

[1]: https://godoc.org/github.com/pkg/errors#Errorf
[2]: https://golang.org/pkg/errors/#New

File: cmd/podman/create.go [hunk 19]
```diff
@@ -116,11 +116,10 @@ func createCmd(c *cli.Context) error {
 		return err
 	}
 
-	options, err := createConfig.GetContainerCreateOptions()
+	options, err := createConfig.GetContainerCreateOptions(runtime)
 	if err != nil {
 		return err
 	}
-
```
Comment (wking) [inline cmd/podman/create.go]: nit: Is this removal intentional?  It looks like unnecessary churn to me.

**Comment by rhatdan** [comment]
Ok I think this is working well with the latest conmon, I have added a test, but it probably will fail in one of these tests if we don't have an updated conmon.  If not then this is ready to merge.

I think this is one of the biggest bugs we have so we need to get this fixed this week.

@mheon @wking @baude @umohnani8 PTAL

**Comment by wking** [comment]
Tests are not happy.  For example, [from the smoketest][1]:

```
======================================================================
ERROR: test_delete_stopped (test.test_containers.TestContainers)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/go/src/github.com/projectatomic/libpod/contrib/python/podman/libs/containers.py", line 155, in stop
    podman.StopContainer(self.id, timeout)
  File "/usr/local/lib/python3.6/site-packages/varlink/client.py", line 75, in _wrapped
    return self._call(method.name, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/varlink/client.py", line 132, in _call
    (message, more) = self._next_varlink_message()
  File "/usr/local/lib/python3.6/site-packages/varlink/client.py", line 102, in _next_varlink_message
    raise e
varlink.error.VarlinkError: {'parameters': {'reason': 'cgroups: unable to remove paths /sys/fs/cgroup/systemd/libpod_parent/libpod-1bf175381336df3725bb06d537796bf613d72ac24ccec5e09a33ae35e1894eb1, /sys/fs/cgroup/freezer/libpod_parent/libpod-1bf175381336df3725bb06d537796bf613d72ac24ccec5e09a33ae35e1894eb1, /sys/fs/cgroup/pids/libpod_parent/libpod-1bf175381336df3725bb06d537796bf613d72ac24ccec5e09a33ae35e1894eb1, /sys/fs/cgroup/net_cls/libpod_parent/libpod-1bf175381336df3725bb06d537796bf613d72ac24ccec5e09a33ae35e1894eb1, /sys/fs/cgroup/net_prio/libpod_parent/libpod-1bf175381336df3725bb06d537796bf613d72ac24ccec5e09a33ae35e1894eb1, /sys/fs/cgroup/perf_event/libpod_parent/libpod-1bf175381336df3725bb06d537796bf613d72ac24ccec5e09a33ae35e1894eb1, /sys/fs/cgroup/cpuset/libpod_parent/libpod-1bf175381336df3725bb06d537796bf613d72ac24ccec5e09a33ae35e1894eb1, /sys/fs/cgroup/cpu/libpod_parent/libpod-1bf175381336df3725bb06d537796bf613d72ac24ccec5e09a33ae35e1894eb1, /sys/fs/cgroup/cpuacct/libpod_parent/libpod-1bf175381336df3725bb06d537796bf613d72ac24ccec5e09a33ae35e1894eb1, /sys/fs/cgroup/memory/libpod_parent/libpod-1bf175381336df3725bb06d537796bf613d72ac24ccec5e09a33ae35e1894eb1, /sys/fs/cgroup/blkio/libpod_parent/libpod-1bf175381336df3725bb06d537796bf613d72ac24ccec5e09a33ae35e1894eb1, /sys/fs/cgroup/devices/libpod_parent/libpod-1bf175381336df3725bb06d537796bf613d72ac24ccec5e09a33ae35e1894eb1, /sys/fs/cgroup/hugetlb/libpod_parent/libpod-1bf175381336df3725bb06d537796bf613d72ac24ccec5e09a33ae35e1894eb1'}, 'error': 'io.projectatomic.podman.ErrorOccurred'}
```

And:

```
[AfterEach] Podman push
  /go/src/github.com/projectatomic/libpod/test/e2e/push_test.go:28
Running: /usr/bin/podman --root /tmp/podman_test549585486/crio --runroot /tmp/podman_test549585486/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --storage-driver=vfs rm -fa
failed to delete container f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b: cgroups: unable to remove paths /sys/fs/cgroup/systemd/libpod_parent/libpod-f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b, /sys/fs/cgroup/freezer/libpod_parent/libpod-f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b, /sys/fs/cgroup/pids/libpod_parent/libpod-f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b, /sys/fs/cgroup/net_cls/libpod_parent/libpod-f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b, /sys/fs/cgroup/net_prio/libpod_parent/libpod-f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b, /sys/fs/cgroup/perf_event/libpod_parent/libpod-f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b, /sys/fs/cgroup/cpuset/libpod_parent/libpod-f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b, /sys/fs/cgroup/cpu/libpod_parent/libpod-f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b, /sys/fs/cgroup/cpuacct/libpod_parent/libpod-f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b, /sys/fs/cgroup/memory/libpod_parent/libpod-f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b, /sys/fs/cgroup/blkio/libpod_parent/libpod-f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b, /sys/fs/cgroup/devices/libpod_parent/libpod-f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b, /sys/fs/cgroup/hugetlb/libpod_parent/libpod-f5dbeaf4bdeb0a87863f27005c001c20f9a727b843e8a0b8fdfcebc8f0223d6b
```

[1]: https://s3.amazonaws.com/aos-ci/ghprb/projectatomic/libpod/0fed8e6300ad8e3c09efeba81b77e0a9a91f6e2e.1.1530040259578152559/output.log

**Comment by wking** [comment]
> Now that I think about it, this makes sense - we're running the cleanup code, which wants to delete the parent of conmon's CGroup, in a program running in said CGroup, so naturally it fails as the CGroup is not empty.

How are the control groups structured?  We should be able to leave any conmon groups alone and only remove *container* groups during cleanup, right?

**Comment by mheon** [comment]
@wking Right now, we create a shared parent for both container and `conmon`. This is what we're removing, as it won't be cleaned up when `conmon` and the container exit.

**Comment by wking** [comment]
> Right now, we create a shared parent for both container and conmon. This is what we're removing, as it won't be cleaned up when conmon and the container exit.

So should we leave the shared parent and conmon groups around for an eventual `delete` call, and only cleanup the container group with the new `cleanup` command?

**Comment by wking** [comment]
> I think this is one of the biggest bugs we have so we need to get this fixed this week.

Do we expect to get [the cgroup structure issue][1] (related to #820?) fixed this week?  If not, can we reroll this to only remove the container cgroup on `cleanup` and leave the `conmon` and parent cgroups alone?  That seems like it would be a fairly narrow pivot that would allow us to cleanup the other stuff at exit-time (although I haven't looked at the code, maybe it's not as narrow a pivot as I think ;).  And it would be one way to get this almost-working PR out the door this week.

[1]: https://github.com/projectatomic/libpod/pull/942#issuecomment-400461771

**Comment by rhatdan** [comment]
If conmon is in a cgroup and it launches podman container cleanup in that cgroup, how can podman container cleanup escape the cgroup?  Is there something that changed in the latest conmon to cause this?  The podman stop $CTR, were working prior to me vendoring in the latest conmon, I believe.
@giuseppe Any ideas?

**Comment by mheon** [comment]
I think a better solution might be to undo our unified CGroup hierarchy work, and move back to a CRI-O esque model where containers and conmons have separate parents. This would actually remove the need for CGroup cleanup entirely, as `runc` at least can be trusted to clean up its own CGroups (I don't know if the spec mandates this, but it seems like a reasonable thing to expect even if not mandated).

**Comment by wking** [comment]
> I think a better solution might be to undo our unified CGroup hierarchy work, and move back to a CRI-O esque model where containers and conmons have separate parents.

I agree, but don't know if we can get that out for this week's release.  Dropping cgroup cleanup here (deferring it to `delete`) seems like something we could get out.

> This would actually remove the need for CGroup cleanup entirely, as runc at least can be trusted to clean up its own CGroups (I don't know if the spec mandates this, but it seems like a reasonable thing to expect even if not mandated).

The spec wording behind that is [here][1].

[1]: https://github.com/opencontainers/runtime-spec/blame/v1.0.1/runtime.md#L133

**Comment by rhatdan** [comment]
@mheon @giuseppe @umohnani8 @wking @baude Cleanup passes.
`podman stop ` No longer cleans up, it relies on the container to clean itself up, which I think is the correct behaviour and fixes some nasty error messages.

**Comment by mheon** [comment]
Alright, this is gross but I guess there's nothing we can do to avoid it for now. LGTM.

File: docs/podman-container-cleanup.1.md [hunk 20]
```diff
@@ -0,0 +1,38 @@
+% podman-container-cleanup "1"
+
+## NAME
+podman\-container\-cleanup - Cleanup Container storage and networks
+
+## SYNOPSIS
+**podman container cleanup [OPTIONS] CONTAINER**
```
Comment (wking) [inline docs/podman-container-cleanup.1.md]: I think you want to distinguish between the literal `podman container cleanup` and the placeholder `OPTIONS` and `CONTAINER`.  From [`man-pages(7)`][1]:

> For commands, this shows the syntax of the command and its arguments (including options); boldface is used for as-is text and italics are used to indicate replaceable arguments. Brackets ([]) surround optional arguments, vertical bars (|) separate choices, and ellipses (...) can be repeated.

So I recommend:

```markdown
+**podman container cleanup** [*options*] *container*
```

[1]: http://man7.org/linux/man-pages/man7/man-pages.7.html

Comment (rhatdan) [reply docs/podman-container-cleanup.1.md]: Ok lets do that in a follow on man page cleanup.  I don't want to hold this up from getting in.

**Comment by rhatdan** [comment]
@rh-atomic-bot r=mheon
